### PR TITLE
Bracket mvar opts agent behaviour

### PR DESCRIPTION
**Pull request by mehakun**
Replace almighty and error-prone `bracketUpdate(A => F[B]):F[B]` with `bracketUpdate(A => F[A]):F[A]` and `bracketModify(A => F[(A, B)]):F[B]`.  
Also fixed bug in Applicative syntax in `unless_` method. It's implementation was same as `when_`

### CODE REVIEW

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 1]
```diff
@@ -36,8 +36,31 @@ object bracket {
...
+      M.flatMap(mvar.take)(oldA =>
+        M.pure(oldA).bracketOpt(use) { (newA, success) => mvar.put(if (success) newA else oldA) }
+      )
+
+    /**
+      * Modify value with effectful transformation, calculating result. In case of error value remains unchanged
+      * @param use function to atomically modify value contained in `MVar` and produce result
+      * @return `F[B]`
+      */
+    // NOTE: maybe return F[(A, B)] to inspect updated mvar?
```
Comment (mehakun) [inline core/src/main/scala/tofu/syntax/bracket.scala:55]: notice the note :)

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 2]
```diff
@@ -36,8 +36,31 @@ object bracket {
...
-      mvar.take.bracketIncomplete(use)(mvar.put)
+
+    /**
+      * Update value with effectful transformation. In case of error value remains unchanged
+      * @param use function to atomically modify value contained in `MVar`
+      * @return `F[A]` modified value contained in `MVar`
+      */
+    def bracketUpdate(use: A => F[A])(implicit FG: Guarantee[F], M: Monad[F]): F[A] =
+      M.flatMap(mvar.take)(oldA =>
+        M.pure(oldA).bracketOpt(use) { (newA, success) => mvar.put(if (success) newA else oldA) }
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:47]: Looks like we need `guaranteeOpt` syntax in the `bracket` package

File: core/src/main/scala/tofu/syntax/bracket.scala (hunk 1 above)
Comment (Odomontois) [reply core/src/main/scala/tofu/syntax/bracket.scala:55]: If user wants some information from `A` they obviously may use appropriate `B`.

**Comment by Odomontois** [comment]
Ok will merge after `NOTE:` deletion

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 3]
```diff
@@ -36,8 +36,31 @@ object bracket {
...
-    def bracketUpdate[B, E](use: A => F[B])(implicit F: Applicative[F], FG: Guarantee[F]): F[B] =
-      mvar.take.bracketIncomplete(use)(mvar.put)
+
+    /**
+      * Update value with effectful transformation. In case of error value remains unchanged
+      * @param use function to atomically modify value contained in `MVar`
+      * @return `F[A]` modified value contained in `MVar`
+      */
+    def bracketUpdate(use: A => F[A])(implicit FG: Guarantee[F], M: Monad[F]): F[A] =
+      M.flatMap(mvar.take)(oldA =>
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:46]: Yeah, that should be FG.bracketOpt(mvar.take)

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 4]
```diff
@@ -28,6 +28,18 @@ object bracket {
     def guaranteeAlways[B](release: F[B])(implicit F: Applicative[F], FG: Guarantee[F]): F[A] =
       FG.bracket(F.unit)(_ => fa)((_, _) => release)
 
+    def guaranteeOpt[B](use: A => F[A])(release: A => F[B])(implicit FG: Guarantee[F], A: Applicative[F]): F[A] =
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:31]: No. The `guaranteeOpt` should be just `(release: Boolean => F[B])`: F[A]

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 5]
```diff
@@ -28,6 +28,18 @@ object bracket {
...
+      FG.bracket(FG.bracket(fa)(use) {
+        case (oldA, success) => release(oldA).whenA(!success)
+      })(newA => release(newA).map(_ => newA)) { (_, _) => A.unit }
+
+    def guaranteeModifyOpt[B, C](
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:36]: I can't understand goal of that.

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 6]
```diff
@@ -12,32 +12,117 @@ import tofu.{Finally, Guarantee}
...
+      * Guarantee finalization of [[fa]].
+      * `release` is always called and boolean indicating success of usage is passed to it
+      * @param release function to release value contained in [[fa]] depending on success of usage
+      * @return `F[A]` updated value
+      */
+    def guaranteeOpt[B](release: Boolean => F[B])(implicit F: Applicative[F], FG: Guarantee[F]): F[A] =
+      FG.bracket(F.unit)(_ => fa)((_, success) => release(success))
+
+    /**
+      * Update value in [[fa]]. In case of error old value of [[fa]] is passed to `commit`, otherwise
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:64]: Well, unsuccessful result is not only the error case, but could be also a cancellation.
I suppose it should be rephrased

File: core/src/main/scala/tofu/syntax/bracket.scala [hunk 7]
```diff
@@ -12,32 +12,117 @@ import tofu.{Finally, Guarantee}
...
+      * result of `use` is passed.
+      * @param use function to modify value contained in [[fa]]
+      * @param commit function to commit passed value
+      * @return `F[A]` updated value
+      */
+    def bracketTransact[B](use: A => F[A])(commit: A => F[B])(implicit FG: Guarantee[F], A: Applicative[F]): F[A] =
```
Comment (Odomontois) [inline core/src/main/scala/tofu/syntax/bracket.scala:70]: This is nice, but I feel the name could be misleading, some people may think it has some relation to STM.
I suggest `bracketTransform` , `bracketReplace` or something like that, emphasising we are dealing with attempt to replace _old_ with _new_

Comment (mehakun) [reply core/src/main/scala/tofu/syntax/bracket.scala:70]: `bracketReplace` sounds good for me. Didn't think about STM...

**Pull request merged**